pennylane_sf/gbs.py

```diff
@@ -0,0 +1,258 @@
+        wires = Wires(wires)
+
+        if self.analytic:
+            return super().probability(wires=wires)
+
+        N = len(wires)
+        samples = np.take(self.samples, self.wires.indices(wires), axis=1)
+
+        probs = all_fock_probs_pnr(samples)
+        ind = np.indices([self.cutoff] * N).reshape(N, -1).T
+
+        probs = OrderedDict((tuple(i), probs[tuple(i)]) for i in ind)
+        return probs
+
+    def jacobian(self, operations, observables, variable_deps):
+        """Calculates the Jacobian of the device.
+
+        Args:
+            operations (list[~pennylane.operation.Operation]): operations to be applied to the
+                device
+            observables (list[~pennylane.operation.Operation]): observables to be measured
+            variable_deps (dict[int, ParameterDependency]): reference dictionary
+                mapping free parameter values to the operations that
+                depend on them
+
+        Returns:
+            array[float]: Jacobian matrix of size (``len(probs)``, ``num_wires``)
+        """
+        self.reset()
+
+        requested_wires = observables[0].wires
+
+        # Create dummy observable to measure probabilities over all wires
+        obs_all_wires = qml.Identity(wires=self.wires)
+        obs_all_wires.return_type = Probability
+        prob = self.execute(operations, [obs_all_wires], parameters=variable_deps)[0]
+
+        jac = self._jacobian_all_wires(prob)
+
+        if requested_wires == self.wires:
+            return jac
+
+        # Unflatten into a [cutoff, cutoff, ..., cutoff, num_params] dimensional tensor
+        jac = jac.reshape([self.cutoff] * self.num_wires + [self.num_wires])
+
+        # Find indices to trace over
+        all_indices = set(range(self.num_wires))
+        requested_indices = set(self.wires.indices(requested_wires))
+        trace_over_indices = all_indices - requested_indices
+
+        traced_jac = np.sum(jac, axis=tuple(trace_over_indices))
+
+        # Flatten into [cutoff ** num_requested_wires, num_params] dimensional tensor
+        traced_jac = traced_jac.reshape(-1, self.num_wires)
+
+        return traced_jac
+
+    def _jacobian_all_wires(self, prob):
+        """Calculates the jacobian of the probability distribution with respect to all wires.
+
+        This function uses Eq. (28) of `this <https://arxiv.org/pdf/2004.04770.pdf>`__ paper.
+
+        Args:
+            prob (array[float]): the probability distribution as a flat array
+
+        Returns:
+            array[float]: the jacobian
+        """
+        jac = np.zeros([len(prob), self.num_wires])
+
+        n = len(self._WAW)
+        disp = np.zeros(2 * n)
+        cov = self.calculate_covariance(self._WAW, hbar=self.hbar)
+        mean_photons_by_mode = photon_number_mean_vector(disp, cov, hbar=self.hbar)
+
+        for i, s in enumerate(np.ndindex(*[self.cutoff] * self.num_wires)):
+            jac[i] = (s - mean_photons_by_mode) * prob[i] / self._params
```

Comment:
Nitpick: I wonder if it is possible to vectorize this?
Something like
```python
basis_states = np.ndindex(*[self.cutoff] * self.num_wires)
indices = np.arange(len(basis_state))
jac = (basis_states - mean_photons_by_mode) * prob / self._params
```
(some modifications would be needed to the shapes to take into account `self._params` and `mean_photons_by_mode`, probably introducing an extra dimension).